rpcserverhandlers_test.go

```diff
@@ -677,81 +690,345 @@ type rpcTest struct {
+	}
+}
+
+// defaultMockAddrManager provides a default mock address manager to be used
+// throughout the tests. Tests can override these defaults by calling
+// defaultMockAddrManager, updating fields as necessary on the returned
+// *testAddrManager, and then setting rpcTest.mockAddrManager as that
+// *testAddrManager.
+func defaultMockAddrManager() *testAddrManager {
+	return &testAddrManager{
+		localAddresses: []addrmgr.LocalAddr{{
+			Address: "127.0.0.184",
+			Port:    uint16(19108),
+			Score:   int32(0),
+		}},
+	}
+}
+
+// defaultMockSyncManager provides a default mock sync manager to be used
+// throughout the tests. Tests can override these defaults by calling
+// defaultMockSyncManager, updating fields as necessary on the returned
+// *testSyncManager, and then setting rpcTest.mockSyncManager as that
+// *testSyncManager.
+func defaultMockSyncManager() *testSyncManager {
+	return &testSyncManager{
+		submitBlock: true,
+		syncHeight:  463074,
+	}
+}
+
+// defaultMockConnManager provides a default mock connection manager to be used
+// throughout the tests. Tests can override these defaults by calling
+// defaultMockConnManager, updating fields as necessary on the returned
+// *testConnManager, and then setting rpcTest.mockConnManager as that
+// *testConnManager.
+func defaultMockConnManager() *testConnManager {
+	testPeer1 := &testPeer{
+		addr:      "127.0.0.210:9108",
+		connected: true,
+		inbound:   true,
+		id:        28,
+	}
+	testPeer2 := &testPeer{
+		addr:      "127.0.0.211:9108",
+		connected: true,
+		inbound:   false,
+		id:        29,
+	}
+	testPeer3 := &testPeer{
+		addr:      "mydomain.org:9108",
+		connected: true,
+		inbound:   false,
+		id:        30,
+	}
+	testPeer4 := &testPeer{
+		addr:      "nonexistentdomain.org:9108",
+		connected: true,
+		inbound:   false,
+		id:        31,
+	}
+	return &testConnManager{
+		connectedCount:   4,
+		netTotalReceived: 9598159,
+		netTotalSent:     4783802,
+		connectedPeers: []rpcserver.Peer{
+			testPeer1,
+			testPeer2,
+			testPeer3,
+			testPeer4,
+		},
+		persistentPeers: []rpcserver.Peer{
+			testPeer1,
+			testPeer2,
+			testPeer3,
+			testPeer4,
+		},
+		addedNodeInfo: []rpcserver.Peer{
+			testPeer1,
+			testPeer2,
+			testPeer3,
+			testPeer4,
+		},
+	}
+}
+
+// defaultMockConfig provides a default rpcserverConfig that is used throughout
+// the tests.  Defaults can be overridden by tests through the rpcTest struct.
+func defaultMockConfig(chainParams *chaincfg.Params) *rpcserverConfig {
+	return &rpcserverConfig{
+		ChainParams:  chainParams,
+		Chain:        defaultMockRPCChain(),
+		AddrManager:  defaultMockAddrManager(),
+		SyncMgr:      defaultMockSyncManager(),
+		ConnMgr:      defaultMockConnManager(),
+		Clock:        &testClock{},
+		TimeSource:   blockchain.NewMedianTime(),
+		Services:     wire.SFNodeNetwork | wire.SFNodeCF,
+		SubsidyCache: standalone.NewSubsidyCache(chainParams),
+	}
+}
 
 func TestHandleAddNode(t *testing.T) {
+	t.Parallel()
+
```

Comment:
Is it necessary to call `t.Parallel()` in all the subtests if it is already called in `*t.Run(name string, f func(t *T)) bool`?

Reply:
This makes the top-level tests run in parallel, whereas the call within `t.Run` makes the subtests (just for that top-level test) run in parallel.

Reply:
So, is the `t.Parallel()` in `t.Run(...)` not really doing anything in this case?

Reply:
It does, without it, the subtests would not run in parallel within the top-level test. Here is a quick example:

With `t.Parallel()` defined only in the top-level tests: the top-level tests will run in parallel but the subtests for each top-level test will be run sequentially.
- `Test1`: `subtest1`
- `Test2`: `subtest1`
- `Test1`: `subtest2`
- `Test2`: `subtest2`
- `Test1`: `subtest3`
- `Test2`: `subtest3`

With `t.Parallel()` defined only in the subtests: the subtests tests will run in parallel but the top-level tests will be run sequentially.
- `Test1`: `subtest3`
- `Test1`: `subtest1`
- `Test1`: `subtest2`
- `Test2`: `subtest2`
- `Test2`: `subtest1`
- `Test2`: `subtest3`

With it defined in both places: both the top-level and subtests will run in parallel.
- `Test2`: `subtest2`
- `Test1`: `subtest3`
- `Test2`: `subtest1`
- `Test1`: `subtest1`
- `Test2`: `subtest3`
- `Test1`: `subtest2`

Reply:
Wow, thanks! I did not get all that from reading the docs.